### PR TITLE
feat: ENT-7918 ENT-7554 Added academies in Algolia index

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
## Description

This PR adds the academy uuids and associated tags to the Algolia index

## Ticket Link

Link to the associated ticket
[ENT-7918](https://2u-internal.atlassian.net/browse/ENT-7918)
[ENT-7554](https://2u-internal.atlassian.net/browse/ENT-7554)

## Sample Index

image

## Post-review

* [ ] Squash commits into discrete sets of changes
* [ ] Ensure that once the changes have been [deployed to stage](https://gocd.tools.edx.org/go/pipeline/activity/stage-enterprise_catalog), prod is manually deployed

### CODE REVIEW

File: enterprise_catalog/apps/api/tasks.py [hunk 1]
```diff
@@ -825,14 +863,20 @@ def _get_algolia_products_for_batch(
         common_catalogs = set()
         if catalog_uuids_for_all_courses_of_program:
             common_catalogs = set.intersection(*catalog_uuids_for_all_courses_of_program)
-        for course_metadata in program_to_courses_mapping[program_content_key]:
```
Comment (irfanuddinahmad) [inline enterprise_catalog/apps/api/tasks.py:828]: This 'for' loop seems like a remnant of a past refactoring ... seems to just add unnecessary cycles

File: enterprise_catalog/apps/api/tasks.py [hunk 2]
```diff
@@ -799,13 +827,23 @@ def _get_algolia_products_for_batch(
         for catalog_query in associated_catalog_queries:
             catalog_queries_by_key[content_key].add((str(catalog_query.uuid), catalog_query.title))
             # This line is possible thanks to `all_catalog_queries` with the prefectch_related() above.
-            associated_catalogs = catalog_query.enterprise_catalogs.all()
+            associated_catalogs = catalog_query.enterprise_catalogs.prefetch_related('academies').all()
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tasks.py]: I think you might want to move this prefetch outside the "first pass".  The way this function was originally designed, all SQL queries are invoked before the first pass (at the beginning of this function), and only O(1) SQL statements.
Can you try updating the `all_catalog_queries` definition (see other comment)?

File: enterprise_catalog/apps/api/tasks.py [hunk 3]
```diff
@@ -737,9 +761,13 @@ def _get_algolia_products_for_batch(
...
+    academy_uuids_by_key = defaultdict(set)
+    academy_tags_by_key = defaultdict(set)
 
     catalog_query_uuid_by_catalog_uuid = defaultdict(set)
     customer_uuid_by_catalog_uuid = defaultdict(set)
+    academy_uuids_by_catalog_uuid = defaultdict(set)
+    academy_tags_by_catalog_uuid = defaultdict(set)
 
     # Create a shared convenience queryset to prefetch catalogs for all metadata lookups below.
     all_catalog_queries = CatalogQuery.objects.prefetch_related('enterprise_catalogs')
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tasks.py]: I think this should become:
```suggestion
    all_catalog_queries = CatalogQuery.objects.prefetch_related('enterprise_catalogs', 'enterprise_catalogs__academies')
```

**Review by pwnage101 — CHANGES_REQUESTED** [review]

File: enterprise_catalog/apps/api/tasks.py (hunk 1 above)
Comment (pwnage101) [reply enterprise_catalog/apps/api/tasks.py:828]: I agree, thanks for catching

File: enterprise_catalog/apps/api/tasks.py (hunk 3 above)
Comment (pwnage101) [reply enterprise_catalog/apps/api/tasks.py]: hopefully this prevents SQL fetches below inside the first pass when you retrieve the academy of a catalog.

File: enterprise_catalog/apps/api/tasks.py (hunk 2 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py]: Done

File: enterprise_catalog/apps/api/tasks.py (hunk 3 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py]: Done

File: enterprise_catalog/apps/api/tests/test_tasks.py [hunk 4]
```diff
@@ -884,7 +899,7 @@ def mock_replace_all_objects(products_iterable):
                 tasks.index_enterprise_catalog_in_algolia_task()  # pylint: disable=no-value-for-parameter
 
         products_found_log_records = [record for record in info_logs.output if ' products found.' in record]
-        assert ' 12 products found.' in products_found_log_records[0]
+        assert ' 14 products found.' in products_found_log_records[0]
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tests/test_tasks.py]: This is not your problem to solve, but can you at least add a note here explaining that the count should really only be "9 products found", but 5 additional algolia products are generated by the test course from self.setUp()?  That extra course is present in the test database and polluting our test numbers whenever the tests don't need it.

File: enterprise_catalog/apps/api/tasks.py [hunk 5]
```diff
@@ -804,8 +835,18 @@ def _get_algolia_products_for_batch(
...
                 # Cache UUIDs related to each catalog.
-                catalog_query_uuid_by_catalog_uuid[str(catalog.uuid)] = (str(catalog_query.uuid), catalog_query.title)
-                customer_uuid_by_catalog_uuid[str(catalog.uuid)] = str(catalog.enterprise_uuid)
+                catalog_query_uuid_by_catalog_uuid[str(catalog.uuid)].add(
+                    (str(catalog_query.uuid), catalog_query.title)
+                )
+                customer_uuid_by_catalog_uuid[str(catalog.uuid)].add(str(catalog.enterprise_uuid))
+                associated_academies = catalog.academies.all()
+                for academy in associated_academies:
+                    associated_academy_tags = academy.tags.all()
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tasks.py:845]: welp, looks like another thing to add to the prefetch: `enterprise_catalogs__academies__tags`

File: enterprise_catalog/apps/api/tasks.py [hunk 6]
```diff
@@ -737,12 +761,19 @@ def _get_algolia_products_for_batch(
...
     customer_uuid_by_catalog_uuid = defaultdict(set)
+    academy_uuids_by_catalog_uuid = defaultdict(set)
+    academy_tags_by_catalog_uuid = defaultdict(set)
 
     # Create a shared convenience queryset to prefetch catalogs for all metadata lookups below.
-    all_catalog_queries = CatalogQuery.objects.prefetch_related('enterprise_catalogs')
+    all_catalog_queries = CatalogQuery.objects.prefetch_related(
+        'enterprise_catalogs',
+        'enterprise_catalogs__academies'
+    )
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tasks.py]:
```suggestion
    all_catalog_queries = CatalogQuery.objects.prefetch_related(
        'enterprise_catalogs',
        'enterprise_catalogs__academies',
        'enterprise_catalogs__academies__tags',
    )
```

I also wonder how to only fetch the columns we need, like tag titles and academy UUIDs.  I spent 5 minutes trying to look that up but couldn't find a solution.

**Review by pwnage101 — CHANGES_REQUESTED** [review]
almost there

File: enterprise_catalog/apps/api/tasks.py (hunk 6 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py]: Not too many columns expected in this Tag model. Also, we only expect to have at most 6-7 academies over the design life of this feature.

File: enterprise_catalog/apps/api/tests/test_tasks.py (hunk 4 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tests/test_tasks.py]: Done

File: enterprise_catalog/apps/api/tasks.py (hunk 5 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py:845]: Done

File: enterprise_catalog/apps/api/tasks.py [hunk 7]
```diff
@@ -737,12 +761,20 @@ def _get_algolia_products_for_batch(
...
     customer_uuid_by_catalog_uuid = defaultdict(set)
+    academy_uuids_by_catalog_uuid = defaultdict(set)
+    academy_tags_by_catalog_uuid = defaultdict(set)
 
     # Create a shared convenience queryset to prefetch catalogs for all metadata lookups below.
-    all_catalog_queries = CatalogQuery.objects.prefetch_related('enterprise_catalogs')
+    all_catalog_queries = CatalogQuery.objects.prefetch_related(
+        'enterprise_catalogs',
+        'enterprise_catalogs__academies',
+        'enterprise_catalogs__academies__tags'
```
Comment (pwnage101) [inline enterprise_catalog/apps/api/tasks.py]: nit: use trailing commas on multi-line argumens

**Review by pwnage101 — APPROVED** [review]

**Pull request merged**